src/cryptonote_core/quorum_cop.cpp

```diff
@@ -0,0 +1,221 @@
+      return;
+
+    uint64_t const execute_justice_from_height = latest_height - loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT;
+    if (height < execute_justice_from_height)
+      return;
+
+    if (height == latest_height)
+    {
+      const std::shared_ptr<quorum_state> latest_quorum = m_core.get_quorum_state(latest_height);
+      if (participates_in_quorum(latest_quorum->nodes_to_test, my_pubkey))
+      {
+        m_core.submit_uptime_proof();
+      }
+    }
+
+    if (m_last_height < execute_justice_from_height)
+      m_last_height = execute_justice_from_height;
+
+    time_t const now = time(nullptr);
+    for (;m_last_height < (height - REORG_SAFETY_BUFFER_IN_BLOCKS); m_last_height++)
+    {
+      const std::shared_ptr<quorum_state> state = m_core.get_quorum_state(m_last_height);
+      if (!state)
+      {
+        // TODO(loki): Fatal error
+        LOG_ERROR("Quorum state for height: " << m_last_height << "was not cached in daemon!");
+        continue;
+      }
+
+      size_t my_index_in_quorum = 0;
+      if (!participates_in_quorum(state->quorum_nodes, my_pubkey, &my_index_in_quorum))
+      {
+        continue;
+      }
+
+      for (size_t node_index = 0; node_index < state->nodes_to_test.size(); ++node_index)
+      {
+        const crypto::public_key &node_key = state->nodes_to_test[node_index];
+        bool vote_off_node = false;
+
+        CRITICAL_REGION_LOCAL(m_lock);
+        auto const it = m_uptime_proof_seen.find(node_key);
+        if (it == m_uptime_proof_seen.end())
+        {
+          vote_off_node = true;
+        }
+        else
+        {
+          size_t time_since_proof = now - it->second;
+          vote_off_node           = (time_since_proof > UPTIME_PROOF_MAX_TIME_IN_SECONDS) ? true : false;
+        }
+
+        if (!vote_off_node)
+          continue;
+
+        loki::service_node_deregister::vote vote = {};
+        vote.block_height        = m_last_height;
+        vote.service_node_index  = node_index;
+        vote.voters_quorum_index = my_index_in_quorum;
+        vote.signature           = loki::service_node_deregister::sign_vote(vote.block_height, vote.service_node_index, my_pubkey, my_seckey);
+
+        cryptonote::vote_verification_context vvc = {};
+        if (!m_core.add_deregister_vote(vote, vvc))
+        {
+          if (vvc.m_invalid_block_height)
+            LOG_ERROR("block height was invalid: " << vote.block_height);
+
+          if (vvc.m_voters_quorum_index_out_of_bounds)
+            LOG_ERROR("voters quorum index specified out of bounds: " << vote.voters_quorum_index);
+
+          if (vvc.m_service_node_index_out_of_bounds)
+            LOG_ERROR("service node index specified out of bounds: " << vote.service_node_index);
+
+          if (vvc.m_signature_not_valid)
+            LOG_ERROR("signature was not valid, was the signature signed properly?");
+        }
+      }
+    }
+
+    prune_uptime_proof_below(execute_justice_from_height);
```

Comment:
I would triggered this during idle in m_core, say once every 30 seconds. Works fine here too, just doesn't feel right because the uptime proof thing is not happening related to the blockchain. This also means the block height stuff in this function is unnecessary - you just have to remove them as long as their timestamp < now - 2 hours and 5 minutes